soc/usbexecd.c

```diff
@@ -0,0 +1,1029 @@
+            if ((pid = fork()) == 0) {
+
+                // For every process in the parent
+                for (int iterator = 0; iterator < N_PROC; iterator++) {
+                    // Get a reference to the correspondong process
+                    procinfo_t* parent_process = processes[iterator];
+
+                    // If this process exists
+                    if (parent_process) {
+                        // Close all of the file descriptors for this child
+                        if ((&parent_process->ctrl)->fd != -1) close((&parent_process->ctrl)->fd);
+                        if ((&parent_process->stdin)->fd != -1) close((&parent_process->stdin)->fd);
+                        if ((&parent_process->stdout)->fd != -1) close((&parent_process->stdout)->fd);
+                        if ((&parent_process->stderr)->fd != -1) close((&parent_process->stderr)->fd);
+                    }
+                }
+
+                // Close socket and fds used by parent
+                close(sock);
+                close(epfd);
+                close(sigfd);
+
+                // Get started on its own task
+                child(ctrl_fd, stdin_fd, stdout_fd, stderr_fd);
+            // If there was an error
+            } else if (pid < 0) {
+                // Notify the user
+                fatal("Error in fork: %s", strerror(errno));
+            // If this is the parent process 
+            } else {
+                // Close the child ends of the pipes
+                close(ctrl_fd);
+                close(stdin_fd);
+                close(stdout_fd);
+                close(stderr_fd);
+                // Set the field for the child process pid
+                p->pid = pid;
+            }
+
+            break;
+
+        case CMD_CLOSE:
+            debug("CMD: Close a process with %d", id);
+            // If the process isn't killed yet
+            if (p->pid) {
+                // Kill it now
+                kill(p->pid, SIGKILL);
+                // Wait for it to be finished
+                waitpid(p->pid, NULL, 0);
+            }
+
+            // Close out all of the pipe buffers if they haven't been closed already
+            if ((&p->ctrl)->fd != -1) pipebuf_out_close(&p->ctrl);
+            if ((&p->stdin)->fd != -1) pipebuf_out_close(&p->stdin);
+            if ((&p->stdout)->fd != -1) pipebuf_in_close(&p->stdout);
+            if ((&p->stderr)->fd != -1) pipebuf_in_close(&p->stderr);
+            // Free the process memory
+            free(processes[id]);
+            // Reset the array entry
+            p = processes[id] = NULL;
+            // Let the sender know that the child process was successfully closed
+            send_header(CMD_CLOSE_ACK, id, 255, 0);
+            // Return
+            break;
+
+        case CMD_KILL:
+            debug("CMD: Kill a process with %d", id);
+            // If this child hasn't already been killed
+            if (p->pid) {
+                // Kill it
```

Comment:
"Send it a signal" to better document that a signal isn't necessarily fatal, despite the name of the syscall used to send one?